ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.hadoop.hive.ql.exec.vector.expressions.gen.FilterLongColEqualDoubleScalar;
+import org.apache.hadoop.hive.ql.exec.vector.expressions.gen.FilterLongColGreaterLongColumn;
+import org.apache.hadoop.hive.ql.exec.vector.util.FakeDataReader;
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.apache.hadoop.hive.ql.optimizer.physical.Vectorizer;
+import org.apache.hadoop.hive.ql.plan.ExprNodeColumnDesc;
+import org.apache.hadoop.hive.ql.plan.FilterDesc;
+import org.apache.hadoop.hive.ql.plan.OperatorDesc;
+import org.apache.hadoop.hive.ql.plan.TopNKeyDesc;
+import org.apache.hadoop.hive.ql.plan.VectorFilterDesc;
+import org.apache.hadoop.hive.ql.plan.VectorTopNKeyDesc;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Tests that are necessary until VectorizedRowBatch selected issue is being fixed:
+ * It has three fields to manage selected elements:
+ * - selected (array)
+ * - selectedInUse
+ * - size (selected size)
+ * <p>
+ * The tricky thing is those fields are all public: so it is possible to modify some of them without keeping the others
+ * in sync.
+ * The issue that is fixed and checked with these tests is when somebody modifies the selected array but forgets to
+ * set the size parameter: Because of that, the Operators can try to process with wrong size value and it can cause
+ * ArrayIndexOutOfBoundsException.
+ * </p>
+ * <p>
+ * Related ticket: HIVE-26992
+ * Those tests can be removed when those fields are public they can be modified with public methods that ensures to
+ * use them properly.
+ * </p>
+ * Ticket to fix the issue: HIVE-26993
+ */
+public class TestVectorOperationProcess {
+
+  HiveConf hiveConf = new HiveConf();
+
+  @Test
+  public void testVectorFilterHasSelectedSmallerThanBatch() throws HiveException {
+    ExprNodeColumnDesc col1Expr = new ExprNodeColumnDesc(Long.class, "col1", "table", false);
+    List<String> columns = new ArrayList<>();
+    columns.add("col1");
+    FilterDesc filterDesc = new FilterDesc();
+    filterDesc.setPredicate(col1Expr);
+    VectorFilterDesc vectorDesc = new VectorFilterDesc();
+
+    Operator<? extends OperatorDesc> filterOp =
+            OperatorFactory.get(new CompilationOpContext(), filterDesc);
+
+    VectorizationContext vc = new VectorizationContext("name", columns);
+
+    VectorFilterOperator vfo = (VectorFilterOperator) Vectorizer.vectorizeFilterOperator(filterOp, vc, vectorDesc);
+
+    vfo.initialize(hiveConf, null);
+    VectorExpression ve1 = new FilterLongColGreaterLongColumn(0,1);
+    VectorExpression ve2 = new FilterLongColEqualDoubleScalar(2, 0);
+    VectorExpression ve3 = new FilterExprAndExpr();
+    ve3.setChildExpressions(new VectorExpression[] {ve1, ve2});
+    vfo.setFilterCondition(ve3);
+
+    FakeDataReader fdr = new FakeDataReader(1024, 3);
+    VectorizedRowBatch vrg = fdr.getNext();
+
+    vrg.selected = new int[] { 1, 2, 3, 4};
+
+    vfo.process(vrg, 0);
```

Comment:
what does it test? should we have any asserts?

Reply:
It is a tricky one. The issue is that without our fix, it should throw an exception. Actually, I didn't find anything that we could assert on. Maybe if we change the fake data generator to not to use random data, we can modify the test to assert if the filtering operator properly filters out the records from the row batch. 
If you check the process method of the VectorizedFilterOperator, it saves it's own state at the beginning of the process method, process the row batch and restores it's original state. So, after the process, there is no difference in the state of the class that we test.

Reply:
maybe at least check that no exceptions are being thrown:
````
Assertions.assertDoesNotThrow(() -> {....})
````

Reply:
Using random data in test is a bad practice. It makes harder to reproduce failures and to find out root cause.
If we really insists adding these tests please use a constant data set and add assertions to the test cases.

Example: IIUC this test case test if the number of selected rows are less than the batch size. In this case an assert like
```
Assert.assertTrue(vrg.selected.length < vrg.size);
```
makes sense.

Please share your thoughts.

Reply:
> Assert.assertTrue(vrg.selected.length < vrg.size);

Both VectorFilterOperator and VectorTopNKeyOperator restores their original state at the end of the process method. So I cannot do assert on that. 
I've tried multiple methods but couldn't figure out a great thing to assert on - yet. 

> Assertions.assertDoesNotThrow(() -> {....})

There is no such a thing in Java4. There is a possibility to annotate the test like this `@Test(expected = None.class)`  but IntelliJ offers me to remove unnecessary code. 

@kasakrisz , @deniskuzZ : what do you think about that?

Reply:
check this https://github.com/apache/hive/blob/6e25a0151d54d2119a2a71e1212f606c8ca55a37/iceberg/iceberg-catalog/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java#L124

Reply:
can't you add the suggested by Kristian check before calling the process?

Reply:
also can't you initialize VectorizedRowBatch by yourself, looks like FakeDataReader was created for some other reasons (simulate actual read ??? there is some sleep functionality)

Reply:
> can't you add the suggested by Kristian check before calling the process?

> Example: IIUC this test case test if the number of selected rows are less than the batch size. In this case an assert like

> Assert.assertTrue(vrg.selected.length < vrg.size);
> makes sense.

Our test case is about manipulating selected array without modifying the size field leads to an inconsistent state so the process method fails with exception. When I modify the selected array and leaves the size field, it is obvious that the selected.length will be smallar than the size.
I give a second chance to Assertions.assertDoesNotThrow